main/src/ca/uwaterloo/flix/language/phase/Reader.scala

```diff
@@ -87,13 +93,32 @@ object Reader {
           val bytes = StreamOps.readAllBytes(zip.getInputStream(entry))
           val str = new String(bytes, flix.defaultCharset)
           val arr = str.toCharArray
-          result += Source(Ast.Input.Text(fullName, str, stable = false), arr, stable = false)
+          result += Source(Ast.Input.PkgTxtFile(p, isPkgSafe), arr, stable = false)
         }
       }
       result.toList
     }.get // TODO Return a Result instead, see https://github.com/flix/flix/issues/3132
   }
 
+  /**
+   * Returns the package's safety signature.
+   */
+  private def isZipPkgSafe(zipFile: ZipFile, zipPath: Path): Boolean = {
+    zipFile.entries().asScala.find(_.getName.endsWith(".toml")) match {
+      case Some(tomlEntry) =>
+        val s = new String(zipFile.getInputStream(tomlEntry).readAllBytes())
+
+        val p = zipPath.toString.stripSuffix(".fpkg") + "/" + tomlEntry.getName
+        ManifestParser.parse(s, Paths.get(p)) match {
+          case Ok(m) => m.safe
```

Comment:
If the manifest is broken,  I think we should complain loudly.

Reply:
I have seen we try to avoid throwing exceptions. In the next PR this will be handled correctly.